src/lib/optimizer/join_graph_edge.hpp

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <iostream>
+#include <memory>
+
+#include "boost/dynamic_bitset.hpp"
+#include "boost/functional/hash.hpp"
+
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+class AbstractExpression;
+
+/**
+ * A bitset that represents a subsets of the vertices of a JoinGraph.
+ *
+ * We use a dynamic bitset here, since it has good support for operations commonly performed on join vertex sets, such
+ * as union, intersection and subtraction.
+ */
+using JoinGraphVertexSet = boost::dynamic_bitset<>;
+
+/**
+ * Represents a (hyper)edge in a JoinGraph.
+ *
+ * Each predicate must operate exactly on the vertices in vertex_set. That is, each predicate must reference columns
+ * from all vertices in vertex_set and no columns from vertices not in vertex_set. If the predicate wouldn't, then it
+ * would belong to another edge.
+ */
+class JoinGraphEdge final {
+ public:
+  explicit JoinGraphEdge(const JoinGraphVertexSet& vertex_set,
+                         const std::vector<std::shared_ptr<AbstractExpression>>& predicates = {});
+
+  void print(std::ostream& stream = std::cout) const;
+
+  JoinGraphVertexSet vertex_set;
+  std::vector<std::shared_ptr<AbstractExpression>> predicates;
```

Comment:
const?

Reply:
Makes it impossible to store JoinGraphEdges in a vector, since they can't be assigned to on vector resize. I grow increasingly hostile towards const-ing. Or the STL. 

Solution would be to const these fields and to store the JoinGraphEdges in a vector of unique_ptrs. Should I do that?

Reply:
Vector of unique_ptrs sounds good. Also, because having a vector of JoinGraphEdges would mean that the copy assignment copies the vector of predicates.

Reply:
I decided that the fields will stay mutual and I will store the edges as instances in a vector. I also made `JoinGraphEdge` a struct, since it is just POD.

*  The `predicates` field needs to be mutable anyway since predicates are collected after the edge is created.
* Copying the occasional vector of predicates isn't ideal, yes, but scattering the edges across the heap (because unique_ptr) is worse. Might just use a `std::list` then, and nobody would do that.